File: src/cupynumeric/sort/sort.cc
/* Copyright 2024 NVIDIA Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#include <cstring>
#include <sstream>

#include "cupynumeric/sort/sort.h"
#include "cupynumeric/sort/sort_cpu.inl"
#include "cupynumeric/sort/sort_template.inl"

#include <functional>
#include <numeric>

namespace cupynumeric {

using namespace legate;

template <Type::Code CODE, int32_t DIM>
struct SortImplBody<VariantKind::CPU, CODE, DIM> {
  using VAL = type_of<CODE>;

  void operator()(TaskContext& context,
                  const legate::PhysicalStore& input_array,
                  legate::PhysicalStore& output_array,
                  const Pitches<DIM - 1>& pitches,
                  const Rect<DIM>& rect,
                  const size_t volume,
                  const size_t segment_size_l,
                  const size_t segment_size_g,
                  const bool argsort,
                  const bool stable,
                  const bool is_index_space,
                  const size_t local_rank,
                  const size_t num_ranks,
                  const size_t num_sort_ranks,
                  const std::vector<comm::Communicator>& comms)
  {
    SortImplBodyCpu<CODE, DIM>()(input_array,
                                 output_array,
                                 pitches,
                                 rect,
                                 volume,
                                 segment_size_l,
                                 segment_size_g,
                                 argsort,
                                 stable,
                                 is_index_space,
                                 local_rank,
                                 num_ranks,
                                 num_sort_ranks,
                                 thrust::host,
                                 comms);
  }
};

/*static*/ void SortTask::cpu_variant(TaskContext context)
{
  sort_template<VariantKind::CPU>(context);
}

namespace  // unnamed
{
static const auto cupynumeric_reg_task_ = []() -> char {
  SortTask::register_variants();
  return 0;
}();
}  // namespace

}  // namespace cupynumeric

File: cupynumeric/_thunk/_sort.py
# Copyright 2024 NVIDIA Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
from __future__ import annotations

from typing import TYPE_CHECKING, cast

from legate.core import get_legate_runtime, types as ty

from ..lib.array_utils import normalize_axis_index
from ..config import CuPyNumericOpCode
from ..runtime import runtime

if TYPE_CHECKING:
    from .._thunk.deferred import DeferredArray


def sort_flattened(
    output: DeferredArray, input: DeferredArray, argsort: bool, stable: bool
) -> None:
    flattened = cast("DeferredArray", input.reshape((input.size,), order="C"))

    # run sort flattened -- return 1D solution
    sort_result = cast(
        "DeferredArray",
        runtime.create_empty_thunk(
            flattened.shape, dtype=output.base.type, inputs=(flattened,)
        ),
    )
    sort_deferred(sort_result, flattened, argsort, stable=stable)
    output.base = sort_result.base
    output.numpy_array = None


def sort_swapped(
    output: DeferredArray,
    input: DeferredArray,
    argsort: bool,
    sort_axis: int,
    stable: bool,
) -> None:
    sort_axis = normalize_axis_index(sort_axis, input.ndim)

    # swap axes
    swapped = input.swapaxes(sort_axis, input.ndim - 1)

    swapped_copy = cast(
        "DeferredArray",
        runtime.create_empty_thunk(
            swapped.shape, dtype=input.base.type, inputs=(input, swapped)
        ),
    )
    swapped_copy.copy(swapped, deep=True)

    # run sort on last axis
    if argsort is True:
        sort_result = cast(
            "DeferredArray",
            runtime.create_empty_thunk(
                swapped_copy.shape,
                dtype=output.base.type,
                inputs=(swapped_copy,),
            ),
        )
        sort_deferred(sort_result, swapped_copy, argsort, stable=stable)
        output.base = sort_result.swapaxes(input.ndim - 1, sort_axis).base
        output.numpy_array = None
    else:
        sort_deferred(swapped_copy, swapped_copy, argsort, stable=stable)
        output.base = swapped_copy.swapaxes(input.ndim - 1, sort_axis).base
        output.numpy_array = None


def sort_task(
    output: DeferredArray, input: DeferredArray, argsort: bool, stable: bool
) -> None:
    legate_runtime = get_legate_runtime()
    task = legate_runtime.create_auto_task(
        output.library, CuPyNumericOpCode.SORT
    )

    uses_unbound_output = runtime.num_procs > 1 and input.ndim == 1

    if runtime.num_gpus > 1:
        task.add_nccl_communicator()
    elif runtime.num_gpus == 0 and runtime.num_procs > 1:
        task.add_cpu_communicator()

    task.add_scalar_arg(argsort, ty.bool_)  # return indices flag
    task.add_scalar_arg(input.base.shape, (ty.int64,))
    task.add_scalar_arg(stable, ty.bool_)
    task.add_input(input.base)

    if uses_unbound_output:
        unbound = runtime.create_unbound_thunk(dtype=output.base.type, ndim=1)
        task.add_output(unbound.base)
        task.execute()
        output.base = unbound.base
        output.numpy_array = None
    else:
        task.add_output(output.base)
        task.add_alignment(output.base, input.base)
        task.execute()


def sort_deferred(
    output: DeferredArray,
    input: DeferredArray,
    argsort: bool,
    axis: int | None = -1,
    stable: bool = False,
) -> None:
    if axis is None and input.ndim > 1:
        sort_flattened(output, input, argsort, stable)
    else:
        if axis is None:
            computed_axis = 0
        else:
            computed_axis = normalize_axis_index(axis, input.ndim)

        if computed_axis == input.ndim - 1:
            sort_task(output, input, argsort, stable)
        else:
            sort_swapped(output, input, argsort, computed_axis, stable)
